contrib/coreos/user-data.example

```diff
@@ -190,58 +209,67 @@ write_files:
+          if [ "$HOST" = "$PUBLIC_IP" ] ; then
+            OSD_ID=$(etcdctl get /deis/store/osds/$PUBLIC_IP)
             break
           fi
-          echo "waiting" && sleep 5
-          TIMEWAITED=$((TIMEWAITED+5))
         done
-        docker stop deis-store-daemon
-        docker exec deis-store-admin ceph osd crush remove osd.$OSD_ID
-        docker exec deis-store-admin ceph auth del osd.$OSD_ID
-        docker exec deis-store-admin ceph osd rm $OSD_ID
-        etcdctl rm /deis/store/osds/$PUBLIC_IP
-        etcdctl rm /deis/store/hosts/$PUBLIC_IP && sleep 10
-        # remove ceph mon
-        docker stop deis-store-monitor || true
-        docker exec deis-store-admin ceph mon remove `hostname -f` # fixme
-        docker stop deis-store-metadata || true
-      fi
+        # if we own an osd and its healthy, try to gracefully remove it
+        if [ ! -z "$OSD_ID" ] && [ ${#OSD_HOSTS[@]} -gt "3" ]; then
+          ADMIN_RUNNING=$(docker inspect --format="{{ .State.Running }}" deis-store-admin)
+          if [ $? -eq 1 ] || [ "$ADMIN_RUNNING" == "false" ]; then
+            echo "deis-store-admin container is required for graceful shutdown"
+            exit 2
+          fi
+          set -e -x -o pipefail
+          CURRENT_STATUS=$(docker exec deis-store-admin ceph health | awk '{print $1}')
+          if [[ "$CURRENT_STATUS" != *"HEALTH_OK"* ]]; then
+            echo "Ceph cluster must be healthy to perform graceful removal"
+            exit 3
+          fi
 
-      # removing the node from etcd
-      NODE=$(etcdctl member list | grep `cat /etc/machine-id` | cut -d ':' -f 1)
-      etcdctl member remove $NODE
+          docker exec deis-store-admin ceph osd out $OSD_ID
+          sleep 30
+          TIMEWAITED=0
+          until [[ $(docker exec deis-store-admin ceph health) == *"HEALTH_OK"* ]]
+          do
+            if [ $TIMEWAITED -gt "1200" ]
+            then
+              echo "ceph graceful removal timeout exceeded"
+              break
+            fi
+            echo "waiting" && sleep 5
+            TIMEWAITED=$((TIMEWAITED+5))
+          done
+          docker stop deis-store-daemon
+          docker exec deis-store-admin ceph osd crush remove osd.$OSD_ID
+          docker exec deis-store-admin ceph auth del osd.$OSD_ID
+          docker exec deis-store-admin ceph osd rm $OSD_ID
+          etcdctl rm /deis/store/osds/$PUBLIC_IP
+          etcdctl rm /deis/store/hosts/$PUBLIC_IP && sleep 10
+          # remove ceph mon
+          docker stop deis-store-monitor || true
+          docker exec deis-store-admin ceph mon remove `hostname -f` # fixme
+          docker stop deis-store-metadata || true
+        fi
+      }
+      etcd_shutdown () {
+        set -e -x -o pipefail
+        # removing the node from etcd
+        NODE=$(etcdctl member list | grep `cat /etc/machine-id` | cut -d ':' -f 1)
```

Comment:
This is not going to work if you use DNS SRV discovery + setting etcd names. I haven't found a reliable way yet to check for that kind of condition

```
etcdctl member list
7c501b0e7a69ecd3: name=node-1 peerURLs=http://node-1.etcd-moo.internal:2380 clientURLs=http://node-1.etcd-moo.internal:2379
9849e1552fbbe9b2: name=node-2 peerURLs=http://node-2.etcd-moo.internal:2380 clientURLs=http://node-2.etcd-moo.internal:2379
b276c45b6b42f883: name=node-3 peerURLs=http://node-3.etcd-moo.internal:2380 clientURLs=http://node-3.etcd-moo.internal:2379
c4d95826ffa38aa9: name=node-4 peerURLs=http://node-4.etcd-moo.internal:2380 clientURLs=http://node-4.etcd-moo.internal:2379
e8c5cab61f1e82b8: name=node-5 peerURLs=http://node-5.etcd-moo.internal:2380 clientURLs=http://node-5.etcd-moo.internal:2379
```

The member id there is not the same as the `fleetctl` `machine-id` so it would only work for `discovery-url` clusters using the `default` name.